op-challenger/fault/solver.go

```diff
@@ -0,0 +1,86 @@
+package fault
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var (
+	ErrNoNextMove = errors.New("no next move")
+)
+
+// Solver uses a [TraceProvider] to determine the moves to make in a dispute game.
+type Solver struct {
+	TraceProvider
+
+	gameDepth int
+}
+
+// NewSolver creates a new [Solver] using the provided [TraceProvider].
+func NewSolver(gameDepth int, traceProvider TraceProvider) *Solver {
+	return &Solver{
+		traceProvider,
+		gameDepth,
+	}
+}
+
+// NextMove returns the next move to make given the current state of the game.
+func (s *Solver) NextMove(claim Claim, parent Claim) (*Response, error) {
+	parentCorrect := s.agreeWithClaim(parent)
+	claimCorrect := s.agreeWithClaim(claim)
+	if parentCorrect && claimCorrect {
+		// We agree with the parent, but the claim is disagreeing with it.
+		// Since we agree with the claim, the difference must be to the right of the claim
+		return s.defend(claim), nil
+	} else if parentCorrect && !claimCorrect {
+		// We agree with the parent, but the claim disagrees with it.
+		// Since we disagree with the claim, the difference must be to the left of the claim
+		return s.attack(claim), nil
+	} else if !parentCorrect && claimCorrect {
+		// Do nothing, we disagree with the parent, but this claim has correctly countered it
+		return s.doNothing(), nil
+	} else if !parentCorrect && !claimCorrect {
+		// We disagree with the parent so want to counter it (which the claim is doing)
+		// but we also disagree with the claim so there must be a difference to the left of claim
+		// Note that we will create the correct counter-claim for parent when it is evaluated, no need to do it here
+		return s.attack(claim), nil
+	}
+	// This should not be reached
+	return nil, ErrNoNextMove
+}
+
+func (s *Solver) doNothing() *Response {
+	return nil
+}
+
+// attack returns a response that attacks the claim.
+func (s *Solver) attack(claim Claim) *Response {
+	claim.Position.Attack()
+	value := s.traceAtPosition(&claim.Position)
+	return &Response{Attack: true, Value: value}
+}
+
+// defend returns a response that defends the claim.
+func (s *Solver) defend(claim Claim) *Response {
+	claim.Position.Defend()
+	value := s.traceAtPosition(&claim.Position)
+	return &Response{Attack: false, Value: value}
+}
+
+// agreeWithClaim returns true if the [Claim] is correct according to the internal [TraceProvider].
+func (s *Solver) agreeWithClaim(claim Claim) bool {
+	ourValue := s.traceAtPosition(&claim.Position)
+	return ourValue == claim.Value
+}
+
+// traceAtPosition returns the [common.Hash] from internal [TraceProvider] at the given [Position].
+func (s *Solver) traceAtPosition(p *Position) common.Hash {
+	index := p.TraceIndex(s.gameDepth)
+	hash, err := s.Get(uint64(index))
```

Comment:
Can update `p.TraceIndex` in a different PR to return a uint64 to keep types consistent.

Reply:
Added a comment note :)